File: rag-endpoint/send-files.py
#!/usr/bin/env python3
"""
Simple example of uploading files to the RAG API.
"""

import argparse
import json
import os
import sys
from datetime import datetime
from pathlib import Path

import requests
from dotenv import load_dotenv
from rich.console import Console
from rich.table import Table


def upload_file(base_url: str, file_path: str, api_key=None):
    """Upload a single file to the RAG system."""
    url = f"{base_url}/files"
    
    headers = {}
    if api_key:
        headers["Authorization"] = f"Bearer {api_key}"
    
    file_path_obj = Path(file_path)
    if not file_path_obj.exists():
        raise FileNotFoundError(f"File not found: {file_path}")
    
    with open(file_path, 'rb') as f:
        # Determine content type based on file extension
        file_ext = file_path_obj.suffix.lower()
        
        content_type = {
            '.pdf': 'application/pdf',
            '.docx': 'application/vnd.openxmlformats-officedocument.wordprocessingml.document',
            '.md': 'text/markdown',
            '.txt': 'text/plain'
        }.get(file_ext, 'application/octet-stream')
        
        filename = file_path_obj.name
        files = {'file': (filename, f, content_type)}
        
        response = requests.post(url, files=files, headers=headers)
        response.raise_for_status()
        
        return response.json()


def check_connection(base_url: str, api_key=None):
    """Simple connection check to the RAG API."""
    headers = {}
    if api_key:
        headers["Authorization"] = f"Bearer {api_key}"
    
    try:
        # Try to connect to the base URL
        response = requests.get(base_url, headers=headers, timeout=5)
        return response.status_code >= 200 and response.status_code < 300, response.status_code
    except requests.exceptions.RequestException as e:
        return False, str(e)


def find_documents(directory_path: str):
    """Find all supported document files in a directory."""
    document_paths = []
    allowed_extensions = {'.pdf', '.docx', '.md', '.txt'}
    
    directory = Path(directory_path)
    if not directory.exists():
        raise FileNotFoundError(f"Directory not found: {directory_path}")
    
    # Walk through files in directory (non-recursive)
    for file_path in directory.iterdir():
        if file_path.is_file() and file_path.suffix.lower() in allowed_extensions:
            document_paths.append(str(file_path.absolute()))
    
    return document_paths


def upload_directory(base_url: str, directory_path: str, api_key=None):
    """Upload all supported files from a directory."""
    console = Console()
    
    document_paths = find_documents(directory_path)
    
    if not document_paths:
        console.print("[yellow]No documents found with supported extensions (.pdf, .docx, .md, .txt)[/yellow]")
        return {}
    
    console.print(f"[green]Found {len(document_paths)} documents[/green]")
    
    # Create a table to track upload progress
    table = Table(title="Document Upload Progress")
    table.add_column("File")
    table.add_column("Status")
    table.add_column("Document ID")
    
    # Create an error table to show full error details
    error_table = Table(title="Upload Errors")
    error_table.add_column("File")
    error_table.add_column("Error Details")
    
    uploaded_docs = {}
    has_errors = False
    
    # Upload each document and record its ID
    for doc_path in document_paths:
        file_name = Path(doc_path).name
        try:
            result = upload_file(base_url, doc_path, api_key)
            doc_id = result.get('id')
            if doc_id:
                uploaded_docs[doc_path] = doc_id
                table.add_row(file_name, "[green]Success[/green]", doc_id)
            else:
                table.add_row(file_name, "[red]Failed[/red]", "N/A")
                error_table.add_row(file_name, "No document ID returned")
                has_errors = True
        except Exception as e:
            table.add_row(file_name, "[red]Failed[/red]", "N/A")
            
            # Generate detailed error message
            error_message = str(e)
            if hasattr(e, 'response') and e.response:
                error_message += f"\nStatus code: {e.response.status_code}"
                try:
                    # Try to parse response as JSON
                    error_json = e.response.json()
                    error_message += f"\nResponse: {json.dumps(error_json, indent=2)}"
                except:
                    # Otherwise show text response
                    error_message += f"\nResponse: {e.response.text}"
            
            error_table.add_row(file_name, error_message)
            has_errors = True
    
    console.print(table)
    
    # Display error table if there were errors
    if has_errors:
        console.print(error_table)
    
    # Generate output filename
    timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
    output_file = f"uploaded_files_{timestamp}.json"
    
    # Only save the document IDs to a file if we successfully uploaded any documents
    if uploaded_docs:
        with open(output_file, 'w') as f:
            json.dump(uploaded_docs, f, indent=2)
        console.print(f"[green]Document IDs saved to {output_file}[/green]")
    else:
        console.print("[yellow]No documents were successfully uploaded, skipping file creation[/yellow]")
    
    return uploaded_docs


def main():
    """Main function to parse arguments and upload files."""
    # Load environment variables from .env file
    load_dotenv()
    
    # Get API URL and key from environment
    rag_api_url = os.getenv("RAG_API_URL")
    rag_api_key = os.getenv("RAG_API_KEY")
    
    # Ensure we have the required environment variables
    if not rag_api_url:
        print("Error: RAG_API_URL not set in environment or .env file")
        return 1
    
    parser = argparse.ArgumentParser(description='Upload files to the RAG API')
    parser.add_argument('--dir', default='./test-data',
                        help='Directory containing files to upload (default: ./test-data)')
    parser.add_argument('--file', '-f', help='Upload a single file instead of a directory')
    
    args = parser.parse_args()
    
    console = Console()
    
    # Check connection to the API first
    is_connected, status = check_connection(rag_api_url, rag_api_key)
    if not is_connected:
        console.print(f"[bold red]Connection to the RAG API failed: {status}[/bold red]")
        return 1
    
    try:
        if args.file:
            # Upload a single file
            try:
                result = upload_file(rag_api_url, args.file, rag_api_key)
                
                # Create a table to show the result
                table = Table(title="File Upload Result")
                table.add_column("File")
                table.add_column("Status")
                table.add_column("Document ID")
                
                file_name = Path(args.file).name
                doc_id = result.get('id')
                if doc_id:
                    table.add_row(file_name, "[green]Success[/green]", doc_id)
                    
                    # Save the document ID to the output file
                    timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
                    output_file = f"uploaded_files_{timestamp}.json"
                    
                    with open(output_file, 'w') as f:
                        json.dump({args.file: doc_id}, f, indent=2)
                    
                    console.print(f"[green]Document ID saved to {output_file}[/green]")
                else:
                    table.add_row(file_name, "[red]Failed[/red]", "N/A")
                
                console.print(table)
                
            except Exception as e:
                console.print(f"[bold red]Error uploading file:[/bold red]")
                
                # Show detailed error information
                error_message = str(e)
                if hasattr(e, 'response') and e.response:
                    console.print(f"Status code: {e.response.status_code}")
                    try:
                        # Try to parse response as JSON
                        error_json = e.response.json()
                        console.print("Response:")
                        console.print_json(json.dumps(error_json))
                    except:
                        # Otherwise show text response
                        console.print(f"Response: {e.response.text}")
                else:
                    console.print(error_message)
                    
                return 1
        else:
            # Upload all files in the directory
            upload_directory(rag_api_url, args.dir, rag_api_key)
    
    except Exception as e:
        console.print(f"[bold red]Error: {str(e)}[/bold red]")
        if hasattr(e, 'response') and e.response:
            console.print(f"Response: {e.response.text}")
        return 1
    
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: rag-endpoint-ts/send-files.ts
#!/usr/bin/env ts-node
/**
 * Simple example of uploading files to the RAG API.
 */

import * as dotenv from 'dotenv';
import axios, { AxiosError, AxiosResponse } from 'axios';
import chalk from 'chalk';
import { Command } from 'commander';
import * as fs from 'fs';
import * as path from 'path';
import Table from 'cli-table3';
import FormData from 'form-data';

interface UploadResult {
  id: string;
  filename: string;
  [key: string]: any;
}

interface UploadedDocs {
  [path: string]: string; // path -> doc_id
}

async function uploadFile(baseUrl: string, filePath: string, apiKey?: string): Promise<UploadResult> {
  const url = `${baseUrl}/files`;
  
  const headers: Record<string, string> = {};
  if (apiKey) {
    headers['Authorization'] = `Bearer ${apiKey}`;
  }
  
  if (!fs.existsSync(filePath)) {
    throw new Error(`File not found: ${filePath}`);
  }
  
  // Create FormData for file upload
  const formData = new FormData();
  
  // Determine content type based on file extension
  const fileExtension = path.extname(filePath).toLowerCase();
  let contentType = 'application/octet-stream'; // Default
  
  switch (fileExtension) {
    case '.pdf':
      contentType = 'application/pdf';
      break;
    case '.docx':
      contentType = 'application/vnd.openxmlformats-officedocument.wordprocessingml.document';
      break;
    case '.md':
      contentType = 'text/markdown';
      break;
    case '.txt':
      contentType = 'text/plain';
      break;
  }
  
  // Append file to form data
  const fileStream = fs.createReadStream(filePath);
  formData.append('file', fileStream, {
    filename: path.basename(filePath),
    contentType: contentType
  });
  
  // Set headers from FormData
  const formHeaders = formData.getHeaders();
  Object.assign(headers, formHeaders);
  
  const response = await axios.post(url, formData, { headers });
  return response.data;
}

async function checkConnection(baseUrl: string, apiKey?: string): Promise<[boolean, string | number]> {
  const headers: Record<string, string> = {};
  if (apiKey) {
    headers['Authorization'] = `Bearer ${apiKey}`;
  }
  
  try {
    // Try to connect to the base URL
    const response = await axios.get(baseUrl, { headers, timeout: 5000 });
    return [response.status >= 200 && response.status < 300, response.status];
  } catch (error: any) {
    return [false, error.message || 'Unknown error'];
  }
}

function findDocuments(directoryPath: string): string[] {
  const documentPaths: string[] = [];
  const allowedExtensions = new Set(['.pdf', '.docx', '.md', '.txt']);
  
  if (!fs.existsSync(directoryPath)) {
    throw new Error(`Directory not found: ${directoryPath}`);
  }
  
  // Walk through files in directory (non-recursive)
  const files = fs.readdirSync(directoryPath);
  for (const file of files) {
    const filePath = path.join(directoryPath, file);
    if (fs.statSync(filePath).isFile() && allowedExtensions.has(path.extname(filePath).toLowerCase())) {
      documentPaths.push(path.resolve(filePath));
    }
  }
  
  return documentPaths;
}

async function uploadDirectory(baseUrl: string, directoryPath: string, apiKey?: string): Promise<UploadedDocs> {
  // Find all documents
  const documentPaths = findDocuments(directoryPath);
  
  if (documentPaths.length === 0) {
    console.log(chalk.yellow('No documents found with supported extensions (.pdf, .docx, .md, .txt)'));
    return {};
  }
  
  console.log(chalk.green(`Found ${documentPaths.length} documents`));
  
  // Create a table to track upload progress
  const uploadTable = new Table({
    head: [
      chalk.white.bold('File'),
      chalk.white.bold('Status'),
      chalk.white.bold('Document ID')
    ],
    style: { head: [], border: [] }
  });
  
  // Create an error table to show full error details
  const errorTable = new Table({
    head: [
      chalk.white.bold('File'),
      chalk.white.bold('Error Details')
    ],
    style: { head: [], border: [] },
    wordWrap: true
  });
  
  const uploadedDocs: UploadedDocs = {};
  let hasErrors = false;
  
  // Upload each document and record its ID
  for (const docPath of documentPaths) {
    const fileName = path.basename(docPath);
    try {
      const result = await uploadFile(baseUrl, docPath, apiKey);
      const docId = result?.id;
      
      if (docId) {
        uploadedDocs[docPath] = docId;
        uploadTable.push([fileName, chalk.green('Success'), docId]);
      } else {
        uploadTable.push([fileName, chalk.red('Failed'), 'N/A']);
        errorTable.push([fileName, 'No document ID returned']);
        hasErrors = true;
      }
    } catch (error: any) {
      uploadTable.push([fileName, chalk.red('Failed'), 'N/A']);
      
      // Generate detailed error message
      let errorMessage = error.message || 'Unknown error';
      if (axios.isAxiosError(error)) {
        const axiosError = error as AxiosError;
        if (axiosError.response) {
          errorMessage += `\nStatus code: ${axiosError.response.status}`;
          try {
            // Try to parse response as JSON
            const errorJson = axiosError.response.data;
            errorMessage += `\nResponse: ${JSON.stringify(errorJson, null, 2)}`;
          } catch {
            // Otherwise show text response
            errorMessage += `\nResponse: ${String(axiosError.response.data)}`;
          }
        }
      }
      
      errorTable.push([fileName, errorMessage]);
      hasErrors = true;
    }
  }
  
  console.log(uploadTable.toString());
  
  // Display error table if there were errors
  if (hasErrors) {
    console.log(errorTable.toString());
  }
  
  // Generate output filename
  const timestamp = new Date().toISOString().replace(/[:.]/g, '').slice(0, 15);
  const outputFile = `uploaded_files_${timestamp}.json`;
  
  // Only save the document IDs to a file if we successfully uploaded any documents
  if (Object.keys(uploadedDocs).length > 0) {
    fs.writeFileSync(outputFile, JSON.stringify(uploadedDocs, null, 2));
    console.log(chalk.green(`Document IDs saved to ${outputFile}`));
  } else {
    console.log(chalk.yellow('No documents were successfully uploaded, skipping file creation'));
  }
  
  return uploadedDocs;
}

async function main(): Promise<number> {
  // Load environment variables from .env file
  dotenv.config();
  
  // Get API URL and key from environment
  const ragApiUrl = process.env.RAG_API_URL;
  const ragApiKey = process.env.RAG_API_KEY;
  
  // Ensure we have the required environment variables
  if (!ragApiUrl) {
    console.error('Error: RAG_API_URL not set in environment or .env file');
    return 1;
  }
  
  // Parse command line arguments
  const program = new Command();
  program
    .option('--dir <directory>', 'Directory containing files to upload', './test-data')
    .option('-f, --file <filepath>', 'Upload a single file instead of a directory')
    .parse(process.argv);
  
  const options = program.opts();
  
  // Check connection to the API first
  const [isConnected, status] = await checkConnection(ragApiUrl, ragApiKey);
  if (!isConnected) {
    console.error(chalk.bold.red(`Connection to the RAG API failed: ${status}`));
    return 1;
  }
  
  try {
    if (options.file) {
      // Upload a single file
      try {
        const result = await uploadFile(ragApiUrl, options.file, ragApiKey);
        
        // Create a table to show the result
        const table = new Table({
          head: [
            chalk.white.bold('File'),
            chalk.white.bold('Status'),
            chalk.white.bold('Document ID')
          ],
          style: { head: [], border: [] }
        });
        
        const fileName = path.basename(options.file);
        const docId = result?.id;
        
        if (docId) {
          table.push([fileName, chalk.green('Success'), docId]);
          
          // Save the document ID to the output file
          const timestamp = new Date().toISOString().replace(/[:.]/g, '').slice(0, 15);
          const outputFile = `uploaded_files_${timestamp}.json`;
          
          fs.writeFileSync(outputFile, JSON.stringify({ [options.file]: docId }, null, 2));
          
          console.log(chalk.green(`Document ID saved to ${outputFile}`));
        } else {
          table.push([fileName, chalk.red('Failed'), 'N/A']);
        }
        
        console.log(table.toString());
      } catch (error: any) {
        console.error(chalk.bold.red('Error uploading file:'));
        
        // Show detailed error information
        if (axios.isAxiosError(error)) {
          const axiosError = error as AxiosError;
          if (axiosError.response) {
            console.error(`Status code: ${axiosError.response.status}`);
            try {
              // Try to parse response as JSON
              const errorJson = axiosError.response.data;
              console.error('Response:');
              console.error(JSON.stringify(errorJson, null, 2));
            } catch {
              // Otherwise show text response
              console.error(`Response: ${String(axiosError.response.data)}`);
            }
          } else {
            console.error(error.message || 'Unknown error');
          }
        } else {
          console.error(error.message || 'Unknown error');
        }
        
        return 1;
      }
    } else {
      // Upload all files in the directory
      await uploadDirectory(ragApiUrl, options.dir, ragApiKey);
    }
  } catch (error: any) {
    console.error(chalk.bold.red(`Error: ${error.message || error}`));
    if (axios.isAxiosError(error)) {
      const axiosError = error as AxiosError;
      if (axiosError.response) {
        console.error(`Response: ${JSON.stringify(axiosError.response.data)}`);
      }
    }
    return 1;
  }
  
  return 0;
}

if (require.main === module) {
  main().then(process.exit).catch(err => {
    console.error(err);
    process.exit(1);
  });
}